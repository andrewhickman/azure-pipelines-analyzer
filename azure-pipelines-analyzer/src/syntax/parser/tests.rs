use insta::assert_debug_snapshot;

use super::Parser;

macro_rules! case {
    ($method:ident($source:expr)) => {{
        let mut parser = Parser::new($source);
        parser.$method();
        let end = parser.pos();
        let parse = parser.finish();
        assert_debug_snapshot!(parse);
        assert_eq!(parse.node.to_string(), $source[..end]);
    }};
}

#[test]
pub fn parse_directive() {
    case!(directive(""));
    case!(directive("foo\nbar\n"));
    case!(directive("%"));
    case!(directive("%  "));
    case!(directive("%\n"));
    case!(directive("%D"));
    case!(directive("%DIR"));
    case!(directive("%DIR\n"));
    case!(directive("%DIR  "));
    case!(directive("%DIR#dir"));
    case!(directive("%DIR#dir\n"));
    case!(directive("%DIR arg1"));
    case!(directive("%DIR arg1#arg"));
    case!(directive("%DIR arg1 #comment"));
    case!(directive("%DIR arg1 ภาษา\n"));
    case!(directive("%DIR arg1\n#comment\n\n"));
    case!(directive("%DIR arg1\r\n#comment\r#"));
}

#[test]
pub fn parse_yaml_directive() {
    case!(directive("%YAML"));
    case!(directive("%YAML\n"));
    case!(directive("%YAML\nfoo"));
    case!(directive("%YAML  "));
    case!(directive("%YAML #comment"));
    case!(directive("%YAML 1.2"));
    case!(directive("%YAML foo.2"));
    case!(directive("%YAML 1 2"));
    case!(directive("%YAML 1.foo"));
    case!(directive("%YAML 100.200"));
    case!(directive("%YAML 100.200#comment"));
    case!(directive("%YAML 100.200 #comment"));
    case!(directive("%YAML 100.200 foo #comment"));
}

#[test]
pub fn parse_tag_directive() {
    case!(directive("%TAG"));
    case!(directive("%TAG\n"));
    case!(directive("%TAG\nfoo"));
    case!(directive("%TAG foo"));
    case!(directive("%TAG foo bar"));
    case!(directive("%TAG foo bar #comment"));
    case!(directive("%TAG !"));
    case!(directive("%TAG !foo"));
    case!(directive("%TAG !foo bar"));
    case!(directive("%TAG !foo!"));
    case!(directive("%TAG !!"));
    case!(directive("%TAG !! foo"));
    case!(directive("%TAG !!!"));
    case!(directive("%TAG ! foo"));
    case!(directive("%TAG !yaml!foo"));
    case!(directive("%TAG !yaml! ภาษา"));
    case!(directive("%TAG !yaml! ,"));
    case!(directive("%TAG !yaml! ["));
    case!(directive("%TAG !yaml! ]"));
    case!(directive("%TAG !yaml! { #comment"));
    case!(directive("%TAG !yaml! } error"));
    case!(directive(
        "%TAG !yaml! !https://example.com:443/;[]()'*~!._,$+=&@?query=foo#fragment"
    ));
    case!(directive("%TAG !yaml! !ภาษา"));
    case!(directive("%TAG !yaml! !https://example.com:443/a%20space"));
    case!(directive("%TAG !yaml! !https://example.com:443/a%__space"));
    case!(directive("%TAG !yaml! tag:yaml.org,2002:"));
    case!(directive("%TAG !yaml! ![example.com]"));
    case!(directive("%TAG !yaml! ![example.com]"));
    case!(directive("%TAG !yaml! !tag:yaml.org,2002:"));
}
