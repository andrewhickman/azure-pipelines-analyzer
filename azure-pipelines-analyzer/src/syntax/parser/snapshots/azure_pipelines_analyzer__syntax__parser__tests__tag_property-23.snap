---
source: azure-pipelines-analyzer/src/syntax/parser/tests.rs
expression: parse
---
Parse {
    node: Root@0..8
      TagProperty@0..8
        SecondaryTagHandle@0..2 "!!"
        TagSuffix@2..8 "suffix"
    ,
    errors: [],
}
